lib/automation.go

```diff
@@ -126,3 +147,183 @@ func (automationImpl) Apply(scope scope, p *ApplyParams) (*ApplyResult, error) {
 	res.RunID = runID
 	return res, nil
 }
+
+// Deploy adds or updates a Dataset, creates or updates an associated Workflow, and, if deployParams.Apply is true, immediately runs the Workflow
+func (automationImpl) Deploy(scope scope, p *DeployParams) error {
+	log.Debugw("deploy", "dataset name", p.Dataset.Name, "peername", p.Dataset.Peername, "workflow id", p.Workflow.ID)
+	if p.Workflow.ID != "" {
+		wf, err := scope.Automation().GetWorkflow(p.Workflow.ID)
+		if err != nil {
+			return nil
+		}
+		if p.Workflow.DatasetID != wf.DatasetID {
+			return fmt.Errorf("deploy: given workflow and workflow on record have different DatasetIDs")
+		}
+		if p.Workflow.OwnerID != wf.OwnerID {
+			return fmt.Errorf("deploy: given workflow and workflow on record have different OwnerIDs")
+		}
+	}
+	go deploy(scope, p)
+	return nil
+}
+
+func deploy(scope scope, p *DeployParams) {
+	vi := dsref.ConvertDatasetToVersionInfo(p.Dataset)
+	ref := vi.SimpleRef().String()
+	deployPayload := event.DeployEvent{
+		Ref:        ref,
+		DatasetID:  p.Dataset.ID,
+		WorkflowID: p.Workflow.ID.String(),
+	}
+	log.Debugw("deploy started", "payload", deployPayload)
+	if scope.Bus() != nil {
+		scope.Bus().PublishID(scope.Context(), event.ETDeployStart, ref, deployPayload)
+	}
+	rollback := true
+	defer func(dp *event.DeployEvent) {
+		if rollback {
+			rp := &RemoveParams{
+				Ref: ref,
+				Revision: &dsref.Rev{
+					Field: "ds",
+					Gen:   1,
+				},
+			}
+			_, err := datasetImpl{}.Remove(scope, rp)
+			if err != nil {
+				log.Debugw("deploy rollback", "error", err)
+			}
+		}
+
+	}(&deployPayload)
+
+	go func(dp *event.DeployEvent) {
+		if err := scope.Bus().PublishID(scope.Context(), event.ETDeploySaveDatasetStart, ref, deployPayload); err != nil {
+			log.Debug(err)
+		}
+	}(&deployPayload)
+
+	saveParams := &SaveParams{
+		Ref:     vi.SimpleRef().String(),
+		Dataset: p.Dataset,
+		Apply:   false,
+	}
+	// TODO(ramfox): bandaid! remove when save can handle having saving a dataset with no body/structure
+	if p.Dataset.ID == "" && p.Dataset.BodyFile() == nil {
+		saveParams.Apply = true
+	}
+	ds, err := datasetImpl{}.Save(scope, saveParams)
+	if err != nil && !errors.Is(err, dsfs.ErrNoChanges) {
+		log.Errorw("deploy save dataset", "error", err)
```

Comment:
use `Debug` level instead. Feels counterintuitive, but CLI users see ugly errors if we leave them at Error level